AD20-16: Adding possibility to execute submitted R analysis in Docker container

hernaldourbina opened the pull request:
[AD20-16](https://odysseusdataservices.atlassian.net/browse/AD20-16): Adding possibility to execute submitted R analysis in Docker container

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/service/impl/DockerServiceImpl.java:
```diff
@@ -0,0 +1,202 @@
+
+            @Override
+            public void close() throws IOException {
+                super.close();
```
Why do you need to override here if you mere call super?

hernaldourbina replied:
In this method of the ResultCallback.Adapter<Frame>, we typically close any resources that need to be explicitly released or cleaned up. At this time, there are no explicit resources that need to be closed yet but I believe it would be good leave it as a placeholder and calling the close method of the superclass (ResultCallback.Adapter<Frame>) to handle any potential cleanup operations defined in the base class and we have also the flexibiity to add our own cleanup logic if necessary in case would be needed.

dmitrys-odysseus replied:
There is flexibility to add the method later on. That will not cause any collateral. I still don't see a reason to put in this override now.

dmitrys-odysseus commented on engine/src/main/resources/application-base.yml:
```diff
@@ -111,4 +111,14 @@ callback:
+
+docker:
+  registry:
+    url:
```
What is the point to add empty values here?

hernaldourbina replied:
I removed the values for pushing the code, I believe a placeholder could work here but the code is prepared to handle these null values.

dmitrys-odysseus replied:
This doesn't answer the question. Any authentication values should come from outside.
What you put in base config should be "meaningful default". I don't see how empty values are meaningful

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/service/impl/DockerServiceImpl.java:
```diff
@@ -0,0 +1,202 @@
+        if (dockerProperties.getRegistryUrl() != null) {
+            builder.withRegistryUrl(dockerProperties.getRegistryUrl());
+        }
+        if (dockerProperties.getRegistryPassword() != null) {
```
Minor: `Optional.ofNullable(dockerProperties.getRegistryUrl()).ifPresent(builder::withRegistryUrl)` plays nicely in such situations

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/api/v1/AnalysisController.java:
```diff
@@ -98,9 +98,11 @@ public AnalysisController(AnalysisService analysisService, CallbackService callb
     public AnalysisRequestStatusDTO analyze(
             @RequestPart("analysisRequest") @Valid AnalysisRequestDTO analysisRequest,
             @RequestPart("file") List<MultipartFile> files,
+            @RequestPart("container") String container,
```
Looks unnecessary

hernaldourbina replied:
This is considered since it was explicitly stated in the ticket: https://odysseusdataservices.atlassian.net/browse/AD20-16 probably for future work.

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/service/impl/AnalysisServiceImpl.java:
```diff
@@ -30,12 +30,7 @@
-import com.odysseusinc.arachne.executionengine.service.DescriptorService;
-import com.odysseusinc.arachne.executionengine.service.RuntimeService;
-import com.odysseusinc.arachne.executionengine.service.SQLService;
+import com.odysseusinc.arachne.executionengine.service.*;
```
Defect: Star import

dmitrys-odysseus commented on commons/src/main/java/com/odysseusinc/arachne/execution_engine_common/descriptor/dto/RuntimeEnvironmentDescriptorsDTO.java:
```diff
@@ -15,11 +18,24 @@ public RuntimeEnvironmentDescriptorsDTO(List<RuntimeEnvironmentDescriptorDTO> de
+        return docker;
+    }
+
+    public void setDocker(boolean docker) {
```
Defect: This is not set anywhere

hernaldourbina replied:
This was added due this comment: "EE api needs to be enchanced to allow datanode to find out if docker is enabled." mentioned in: https://bitbucket.org/Odysseus/ods-execution-engine/pull-requests/13#comment-453894355. Please see note.

dmitrys-odysseus replied:
This does not address the defect. The value is not set and thus will not be sent back to datanode

dmitrys-odysseus commented:
Important:  Please updated README to include a sample request to perform docker execution that can be easily copy-pasted for testing.

dmitrys-odysseus commented:
Defect: Missing timeout handling

dmitrys-odysseus commented:
Defect: In my testing, the current implementation didn't pull an image missing locally

hernaldourbina commented:
> Important: Please update README to include a sample request to perform docker execution that can be easily copy-pasted for testing.

I have updated the README in order to provide guidance for testing this implementation locally. In addition, I have addressed and answered the additional review comments. Thanks for the feedback. Thanks for the feedback.

dmitrys-odysseus reviewed (changes requested):
There are 2 blocking issues at the moment:
1. if image is specified in AnalysisSyncRequestDTO.dockerImage - it has to be pulled and used for execution. 
2. The environment configuration is not passed over to docker execution. See RService.buildRuntimeEnvVariables() for the details

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java:
```diff
@@ -0,0 +1,268 @@
+        log.info("Starting container: {}", containerId);
+        dockerClient.startContainerCmd(containerId).exec();
+
+        String codeDir = "/bin";
```
Why `/bin`? Shouldn't we have dedicated directory for analysis? Or at least `/etc/smth`

hernaldourbina replied:
`/bin` is always present in all images while other folder would need to be created or an error could occur.

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java:
```diff
@@ -0,0 +1,268 @@
+                return new ExecutionOutcome(Stage.COMPLETED, e.getMessage(), "");
+            }
+        }, executor);
+        return new SQLService.SqlOverseer(analysis.getId(), started, stdout, future);
```
Should be DockerOverseer, I assume

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java:
```diff
@@ -0,0 +1,268 @@
+            public void onNext(Frame item) {
+                log.info(item.toString(), analysis);
+                try {
+                    String output = new String(item.getPayload());
```
Defect: missing encoding

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java:
```diff
@@ -0,0 +1,268 @@
+    private ResultCallback.Adapter<Frame> getAdapter(AnalysisSyncRequestDTO analysis, BiConsumer<String, String> callback) {
+        return new ResultCallback.Adapter<Frame>() {
+            @Override
+            public void onNext(Frame item) {
```
I think docker Frame is typically a single log line. These shouldn't be sent individually to datanode, however BiConsumer that takes stage as a separate argument is designed for a frame that will be sent immediately. So using it here looks like an abstraction failure.

hernaldourbina replied:
The intention here is to forward the processed output to datanode in real time as it streams output from the Docker container.

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java:
```diff
@@ -0,0 +1,268 @@
+                    String output = new String(item.getPayload());
+                    log.info("Execution output: {}", output);
+                    callback.accept(Stage.EXECUTE, output);
+                } catch (Exception e) {
```
What kind of exception you are expecting here?

hernaldourbina replied:
It would be `RuntimeException`.

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java:
```diff
@@ -0,0 +1,268 @@
+                log.error("Error during execution", throwable);
+                String error = throwable.getMessage();
+                log.error("Execution error: {}", error);
+                callback.accept(Stage.EXECUTE, error);
```
1. Handling an exception here, why `Stage.Execute` is being used?
2. Since both stdout and stderr are mixed, I suggest we add some prefix for error frames.

hernaldourbina replied:
By default, docker/container logs show the command’s STDOUT and STDERR. But addressing to prefix stderr frames.

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java:
```diff
@@ -0,0 +1,268 @@
+            public void onError(Throwable throwable) {
+                log.error("Error during execution", throwable);
+                String error = throwable.getMessage();
+                log.error("Execution error: {}", error);
```
Logging the same error twice?

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java:
```diff
@@ -0,0 +1,268 @@
+                log.info("Container created for image {} with container id {}", image, container.getId());
+                String containerId = container.getId();
+                runContainer(client, containerId, file, analysis, callbackWithStdOut);
+                remove(client, containerId);
```
If exception occurs in runContainer, the container will not be removed. Is this intentional?

hernaldourbina replied:
Thanks. This is not intentional. I moved this code to the finally block inside `runContainer` to ensure resources are always cleaned up, even in case of exceptions but it could be useful for debugging purposes.

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java:
```diff
@@ -0,0 +1,268 @@
+
+            @Override
+            public void close() throws IOException {
+                super.close();
```
Unnecessary override.

hernaldourbina replied:
There are no explicit resources that need to be closed yet but I believe it would be good leave it as a placeholder for adding our own cleanup logic if necessary later.

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java:
```diff
@@ -0,0 +1,268 @@
+
+        @Override
+        public AnalysisRequestTypeDTO getType() {
+            return SQL;
```
Defect: definitely not `SQL` here. Discuss, whether it should be `R` or new type. @alex-odysseus do you have an opinion on this?

hernaldourbina replied:
Changing it to `R`, while confirming. A new Type would be good.

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java:
```diff
@@ -0,0 +1,268 @@
+        Optional.ofNullable(dockerProperties.getRegistryUsername()).ifPresent(builder::withRegistryUsername);
+        DefaultDockerClientConfig config = builder.build();
+
+        URI host = Optional.ofNullable(System.getProperty("use.socket.path")).map(path ->
```
We configure application using environment variables. What is the reason to choose system property for this option?

hernaldourbina replied:
Mostly related to maintains consistency.

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java:
```diff
@@ -0,0 +1,268 @@
+        @Override
+        public CompletableFuture<ExecutionOutcome> abort() {
+            return result.isDone() ? result : CompletableFuture.completedFuture(
+                    new ExecutionOutcome(Stage.ABORT, "Abort is not supported for SQL analysis", null)
```
SQL again

hernaldourbina commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/RService.java:
```diff
@@ -24,8 +23,10 @@
+import org.springframework.stereotype.Service;
 
 @Slf4j
+@Service
```
You're right. I might have missed this was an abstract class. I will add the `@Service` annotation to `DockerService `class since `Autowired `members should be defined in a Spring Bean.

dmitrys-odysseus replied:
This change has broken the application:
```
Caused by: java.lang.IllegalStateException: Duplicate key com.odysseusinc.arachne.executionengine.execution.r.DockerService@58cf8f94
	at java.util.stream.Collectors.lambda$throwingMerger$0(Collectors.java:133) ~[na:1.8.0_392]
	at java.util.HashMap.merge(HashMap.java:1255) ~[na:1.8.0_392]
	at java.util.stream.Collectors.lambda$toMap$58(Collectors.java:1320) ~[na:1.8.0_392]
	at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169) ~[na:1.8.0_392]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384) ~[na:1.8.0_392]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482) ~[na:1.8.0_392]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472) ~[na:1.8.0_392]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) ~[na:1.8.0_392]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:1.8.0_392]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:566) ~[na:1.8.0_392]
	at com.odysseusinc.arachne.executionengine.execution.AnalysisService.<init>(AnalysisService.java:92) ~[30d8ab8a-5585-4ee9-959c-26a69c5d5890/:na]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.8.0_392]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:1.8.0_392]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:1.8.0_392]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[na:1.8.0_392]
```

hernaldourbina replied:
I will remove this change since looks like it was already registered as a Bean.

hernaldourbina commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java:
```diff
@@ -0,0 +1,268 @@
+                .exec();
+    }
+
+    public void pullImageIfNotExists(String imageName) {
```
Thanks. This method was changed to ensure expected behaviour.

hernaldourbina commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java:
```diff
@@ -0,0 +1,268 @@
+
+    protected Map<String, String> buildRuntimeEnvVariables(DataSourceUnsecuredDTO dataSource, Map<String, String> krbProps) {
+        Map<String, String> environment = super.buildRuntimeEnvVariables(dataSource, krbProps);
+        environment.put(DOCKER_ENV_MEMORY_LIMIT_KEY, DOCKER_ENV_MEMORY_LIMIT_VALUE);
```
I added these docker properties as a sample, but I believe they could be safely removed and just pass the runtime env variables into the docker execution.

dmitrys-odysseus replied:
I'm confused, this is production source code. Why would a "sample" be here?

hernaldourbina replied:
Yeah, additional docker configuration was not needed and removed. I also added the code to pass the runtime env vars.

dmitrys-odysseus commented on engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java:
```diff
@@ -0,0 +1,230 @@
+                .withRemotePath(codeDir).exec();
+
+        String baseDir = codeDir + "/" + analysisDir.getName() + "/";
+        String workingDir = baseDir + (execFileName.split("/").length > 1 ? execFileName.split("/")[0] : "");
```
I don't understand the intent of string manipulation here, could you please clarify it?

hernaldourbina replied:
Sure, this code dynamically constructs the directory path inside the Docker container where the R script is located and will be executed. Also consider the possibility of the script being in a subdirectory within the analysis directory.

dmitrys-odysseus commented on engine/src/main/resources/application-base.yml:
```diff
@@ -111,4 +111,14 @@ callback:
+    url: https://registry-1.docker.io
+    username:
+    password:
+use:
```
"use" is a weird prefix. I suggest to put all the docker configuration under 'docker' prefix. e.g.:
`use.docker` -> `docker.enabled`
`use.socket.path` -> `docker.socket`
